pkg/proxy/kafka.go

```diff
@@ -69,45 +161,54 @@ func createKafkaRedirect(r *Redirect, conf kafkaConfiguration, endpointInfoRegis
 		redir.conf.lookupNewDest = lookupNewDest
 	}
 
-	marker := 0
-	if !conf.noMarker {
-		markIdentity := int(0)
-		// As ingress proxy, all replies to incoming requests must have the
-		// identity of the endpoint we are proxying for
-		if r.ingress {
-			markIdentity = int(r.localEndpoint.GetIdentity())
-		}
-
-		marker = getMagicMark(r.ingress, markIdentity)
-	}
-
-	// Listen needs to be in the synchronous part of this function to ensure that
-	// the proxy port is never refusing connections.
-	socket, err := listenSocket(fmt.Sprintf(":%d", r.ProxyPort), marker)
-	if err != nil {
-		return nil, err
+	key := mapKey(r.dstPort, r.ingress, r.endpointID)
+	log.Debugf("Registering with port: %d, ingress: %v, EID: %d", r.dstPort, r.ingress, r.endpointID)
+	mutex.Lock()
+	if _, ok := kafkaRedirects[key]; ok {
+		mutex.Unlock()
+		panic("Kafka redirect already exists for the given dst port and endpoint ID")
```

Comment:
let's not panic here since, as far as I understand, this `createKafkaRedirect` can be called by the same `r.dstPort, r.ingress, r.endpointID` tuple at the "same time". For example 2 redirects happening at the same time, the first redirect calls this function which sets this up, the 2nd redirect also calls this function and will wait on line 166 until the mutex is released by the 1rst redirect setup. Once the mutex is released the 2nd redirect will panic because the 1st redirect had already configure this

Reply:
Higher level mutex in `CreateOrUpdateRedirect()` protects against this from happening.